server/routerlicious/packages/services-utils/src/documentKeyRetriever.ts

```diff
@@ -0,0 +1,132 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IDocumentManager } from "@fluidframework/server-services-core";
+import { Lumberjack } from "@fluidframework/server-services-telemetry";
+
+export class DocumentKeyRetriever {
+	public constructor(
+		/* The type of the redis variable is currently set to "any," which is not ideal. The reason for this
+		 * is that there is currently a different RedisCache and ICache implementation for Routerlicious, Gitrest, and Historian, which all differ slightly.
+		 * The RedisCache intented to be passed into this constructor is the one from Historian, however I am not able to import Historian's ICache or RedisCache into r11s.
+		 * Using r11s' ICache also doesn't work because it has different return types than the Historian ICache, which has templated outputs/inputs for cache.get and cache.set.
+		 *
+		 * Here is what I propose:
+		 * 1. We leave it as type "any" for now
+		 * 2. During Microsoft's FHL week (September 11-15), I will work on refactoring RedisCache such that Routerlicious, Gitrest, and Historian will all use the same ICache/RedisCache (with the prefix being passed into the constructor)
+		 */
+		private readonly redis: any,
+		private readonly documentManager: IDocumentManager,
+		private readonly loggingEnabled: boolean = true,
+	) {
+		if (!redis) {
+			throw new Error("Redis cache is undefined.");
+		}
+		if (!documentManager) {
+			throw new Error("Document manager is undefined.");
+		}
+	}
+
+	public async getKeyCosmos<ValType>(
+		keyName: string,
+		tenantId: string,
+		documentId: string,
+		useCosmosCache: boolean = true,
```

Comment:
Nit: Would it be better to call this useCache instead? When I first read useCosmosCache, I thought we were using this flag to say we want to use cosmos db as the cache.

Reply:
I included cosmos in that parameter name because it is also used in the getKeyRedisFallback method, and I didn't want it to seem like the flag is whether or not you should use the redis cache. I see what you mean though, do you think one of these other names would be better?:
- useCachedDocumentDetails
- useCachedDocDetails
- useCachedDocument
- useCachedCosmos

Reply:
Switched to useCachedDocument and pushed